Cmdline refactoring

andreitraistaru opened the pull request:
### Summary of the PR

Reworking the Cmdline to leverage handling of boot args and init args. This PR should solve the #116 issue.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [ ] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [ ] All added/changed functionality has a corresponding unit/integration
  test.
- [ ] Any newly added `unsafe` code is properly documented.

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -116,134 +97,98 @@ impl Cmdline {
         Cmdline {
-            line: String::with_capacity(capacity),
-            capacity,
+            boot_args: String::with_capacity(cmdline_capacity),
+            init_args: String::with_capacity(cmdline_capacity),
```
This allows both components of the command line to have the maximum capacity, which would violate the capacity constraint on the entire `Cmdline` object. I don't think we can preallocate space for `String`s here anymore.

andreeaflorescu replied:
If we remove the cmdline_capacity, the code is going to be greatly simplified. We could have a `MAX_SIZE` to make sure that the cmdline is still within reason to protect against potential malicious actors.

andreitraistaru replied:
I though about preallocating cmdline_capacity for both boot_args and init_args to simplify the implementation. Implementing like this guarantees that we have enough space even though no init arg will be provided (and thus, the entire space will be used for boot_args). We will NOT exceed cmdline_capacity when joining boot_args and init_args because before each insert in the Cmdline, we check the condition for capacity of the joined cmdline (the function has_capacity checks for the sum of lengths of boot_args and init_args). Yes, we can definitely implement without preallocations and keep the same checks - the only disadvantage that I can see for this approach would be the overhead introduced by repetitive inserts into string, but probably this would be insignificant as long as there are few inserts for the Cmdline. I wouldn't remove the cmdline_capacity from the constructor parameters as long as this is a constant specific to the architecture and I consider it part of the abstractions the rust-vmm/linux-loader imposes.

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -72,18 +61,6 @@ fn valid_str(s: &str) -> Result<()> {
     }
 }
 
-fn valid_element(s: &str) -> Result<()> {
```
Why has this been removed? Components of the boot args (keys, corresponding values and standalone boot args) should still be validated.

andreitraistaru replied:
it was removed because this method was used only inside the methods removed (and mentioned above). Considering the removal of these methods the valid_element was no longer needed. Considering the comments above, I'll rollback this method as well.

aghecenco commented on src/cmdline/mod.rs:
````diff
@@ -116,134 +97,98 @@ impl Cmdline {
     /// ```
-    pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
-        let s = slug.as_ref();
+    pub fn insert_init_arg<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
````
This function looks a lot like the current body of `insert_boot_arg`, which in turn looks somewhat like the old `insert_str`, so I would keep 1 single function for inserting a plain string alongside key-value insertion functions for boot args only.

andreitraistaru replied:
Ok, so if I got it right you suggest to keep 2 functions: one for inserting a string into init_args and the other one for inserting a key-value pair into boot_args. To make it clear for the user, I would suggest to have a naming like `insert_init_args_str()` and `insert_boot_args_pair()`. Do you agree with this naming convention? Otherwise, it would be a little bit confusing for the user, even though the types of the parameters of the functions clarifies the ambiguity.

andreitraistaru replied:
Also, I consider useful to keep a function that inserts a string into boot args as long as we might need to append boot args that are not key-value pairs.

andreeaflorescu replied:
Since we already have breaking changes at the interface level, I was thinking about making a more clear separation of concerns between bootargs and initargs by creating 2 separate structures to handle each. For the operations that are the same for both of them, we can re-use the code by implementing the same trait for both structures. For the functionality that is different, we need to define it in the corresponding structure. One example for separate functionality is the `insert_virtio` function which needs to be implemented only for `BootArgs`.

This will result in more code, but I think it makes the interface more logical as well.

```rust
pub struct BootArgs(String);
pub struct InitArgs(String);

pub trait UpdateArgs {
    fn insert_key_value<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()>;
    fn insert_multiple<T: AsRef<str>>(&mut self, key: T, vals: &[T]) -> Result<()>;
    ......
    /// we need to add all the other common functions here as well.
}

/// To reuse the code we can implement the trait for String and then just have blanket implementations for both InitArgs and BootArgs.
/// There are probably better ways to auto-implement the trait, but we need to think about this a bit more.

#[derive(Clone, Debug)]
pub struct Cmdline {
    boot_args: BootArgs,
    init_args: InitArgs,
}

impl UpdateArgs for BootArgs {...}
impl UpdateArgs for InitArgs {...}

impl BootArgs {
    pub fn insert_virtio_device.....
    /// here we need to add any other functions that are specific to the boot args.
.....
}

impl Cmdline {
     pub fn boot_args_mut(&mut self) -> &mut BootArgs {....}
     pub fn init_args_mut(&mut self) -> &mut InitArgs {...}
}

......
fn how_to_use() {
   let mut cmdline = Cmdline::try_from("some string");
   cmdline.boot_args_mut().insert_virtio_device(.....);
   cmdline.boot_args_mut().insert_multiple(.....);
}



```

andreitraistaru replied:
Working on this

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -336,101 +287,88 @@ impl Cmdline {
+    fn try_from(
+        (cmdline_capacity, cmdline_raw): (usize, String),
+    ) -> result::Result<Self, Self::Error> {
+        let (boot_args, init_args) = cmdline_raw.split_once("-- ").unwrap();
```
Don't `unwrap` in a `TryFrom`, propagate the error instead.

andreitraistaru replied:
Ok, it makes sense... as long as the method's name starts with 'try'.

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -336,101 +287,88 @@ impl Cmdline {
+        let (boot_args, init_args) = cmdline_raw.split_once("-- ").unwrap();
+
+        if boot_args.len() + init_args.len() + CMDLINE_NO_OF_AUXILIARY_CHARS > cmdline_capacity {
+            return Err(());
```
This should be a specific error, not `()`.

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -336,101 +287,88 @@ impl Cmdline {
-        let s = CString::new(cl).expect("failed to create CString from Cmdline");
-        assert_eq!(s, CString::new("hello=world").unwrap());
-    }
+    const DEFAULT_CMDLINE_RAW : &str = "console=ttyS0 nomodules random.trust_cpu=on reboot=k panic=1 pci=off ip=:::::eth0:dhcp init=/bin/cat -- /etc/password";
```
This is the same as `DEFAULT_CMDLINE_BOOT_ARGS` plus an init arg so I think it should be defined as such.

aghecenco replied:
I don't think the *default* command line we test with needs to be this explicit. Several of these boot args bring no additional value - from a testing standpoint, all key-value args are the same.

aghecenco reviewed (dismissed):
There are some breaking changes I disagree with, and I left comments throughout the code to explain. I also have some style nits but those can wait until later.
I also have some commit-specific comments:
- Commit 1caf1a7f4b15ea57e9f751786544f6ffe2ad63af :  typo `abe/able`
- Commit 9dbaaf5c6e2a63991730022bc6a19247332b2929 should not exist. It looks like its purpose is to test integration with a consumer of this crate. Such changes are dev-only and should be made in the vmm that consumes this crate, not the other way around.
- Commit c52ba27f801bed35faff9d51f58bc30b63588f61 should not exist either, tests should pass on each commit.

aghecenco reviewed (commented):
@andreitraistaru  additionally, as this PR fixes a limitation (essentially a bug) - namely, that `Cmdline` can't handle init args -, I think you should open an issue for that and assign the PR to it.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -116,134 +97,98 @@ impl Cmdline {
-    pub fn new(capacity: usize) -> Cmdline {
-        assert_ne!(capacity, 0);
+    pub fn new(cmdline_capacity: usize) -> Cmdline {
+        assert!(cmdline_capacity > CMDLINE_NO_OF_AUXILIARY_CHARS);
```
Generally speaking we should never use asserts in library code for invalid input. Asserts in library should be reserved for programming errors. In this case, this is a user error for which we should return an error.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -254,12 +199,18 @@ impl Cmdline {
+    pub fn as_str(&self) -> String {
+        if self.init_args.is_empty() {
+            self.boot_args.to_string()
+        } else if self.boot_args.is_empty() {
+            format!("-- {}", self.init_args)
```
Would this be a valid kernel command line? I wonder if we should return an error in this case.

andreitraistaru replied:
Regarding String::with_capacity(cmdline_capacity) I propose to declare it as String::new() and not provide any initial capacity at all. Before any insertion, we perform capacity checks considering both boot args and init args (if exists) such that insertion of too large args to be impossible. Do you have any concern with this approach?

andreitraistaru replied:
Regarding String::with_capacity(cmdline_capacity) I propose to declare it as String::new() and not provide any initial capacity at all. Before any insertion, we perform capacity checks considering both boot args and init args (if exists) such that insertion of too large args to be impossible. Do you have any concern with this approach?

andreitraistaru replied:
Regarding the usecase in which we return format!("-- {}", self.init_args), I don't see it as an error because the method is called as_str so it does not imply any validity checks done by the cmdline; it just implies a representation of the current state of the cmdline; furthermore, we do not perform any validity checks of the boot arguments the user can insert into the cmdline either (apart from the basic check to make sure it's an ASCII string and it doesn't contain any space); thus, it's the responsibility of the user to provide correct data, we just offer the data structure for implementing a Cmdline.

andreeaflorescu replied:
The meaning of `as_str` is that you can pass this to the kernel, so in that sense the `Cmdline` will not get any more pushes. We are working with a fail fast methodology. That means that if there are error cases that we can check ourselves in a reasonable way, we should do that. For example, that is why are checking for ASCII only strings. We could also not check, and let the Linux Kernel fail to boot. But that would mean that we are passing erroneous data down to the lower levels for no good reason. Hence, my question whether Linux accepts a command line that only has init args.

andreitraistaru replied:
Yes, you are right somehow it's always better to catch errors as early as possible. I'll investigate this usecase and come back with a result.

andreeaflorescu reviewed (commented):
There are some high level changes required to the PR, I did not review the tests and in depth the TryFrom functionality as they're likely to change once a newer revision is pushed.

andreitraistaru commented:
Thank you so much for your implication and help! I'll take each comment and address it accordingly

andreitraistaru commented on src/cmdline/mod.rs:
````diff
@@ -116,134 +97,98 @@ impl Cmdline {
-    /// assert_eq!(cl_cstring.to_str().unwrap(), "foo=bar");
+    /// assert_eq!(cl_cstring.to_str().unwrap(), "foobar");
     /// ```
-    pub fn insert<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()> {
````
Ok, that was one aspect I was not sure about. I saw that we don't use it anywhere in Firecracker so I decided to remove it. However, from the abstraction point of view, yes... you are right... it would be better to keep this feature available. I'll add the methods related to key/value back into the commit.

andreeaflorescu replied:
This is not only used by Firecracker, but other customers as well. We shouldn't remove existing functionality.

andreitraistaru commented on src/cmdline/mod.rs:
````diff
@@ -116,134 +97,98 @@ impl Cmdline {
-    /// let cl_cstring = CString::new(cl).unwrap();
-    /// assert_eq!(cl_cstring.to_str().unwrap(), "foo=bar,baz");
-    /// ```
-    pub fn insert_multiple<T: AsRef<str>>(&mut self, key: T, vals: &[T]) -> Result<()> {
````
yes, I'll rollback this method as well.

andreeaflorescu commented:
@andreitraistaru why are you removing the coverage files?

andreitraistaru commented:
@andreeaflorescu sorry, I mistakenly deleted those files.... rollbacking them now. thanks

andreitraistaru commented:
I did the separation of concerns with BootArgs and InitArgs as discussed previously; now, going back to write some documentation and fix the tests to have all green in buildkite; I'll let you know when ready for a new review.

There is one concern from my point of view: the way we currently handle cmdline capacity checks; I implemented a simple check when inserting in BootArgs and InitArgs structs such that user will not be able to abuse the memory and a final check in the as_str() method where the total capacity should be less or equal to the capacity given when instantiating a Cmdline object. Is it ok like this? Did I miss something? It would have been way too complicated to check the total length of the Cmdline when inserting into a BootArgs struct so I consider that the way I implemented is a good compromise

andreitraistaru commented:
I did some code changes to handle double quotes handling as well.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -336,172 +250,641 @@ impl Cmdline {
 }
 
+impl UpdateArgs for BootArgs {
+    fn has_capacity(&self, more: usize) -> bool {
```
Why is this implemented for each type of Arg? We should be able to use the same implementation twice. For example, we could implement the trait for String, and then here just call `self.0.has_capacity()`. Same for InitArgs. This also applies to the other functions that are duplicated.

andreitraistaru replied:
yes, you are right.... still not get used to rust .... I initially tried to move has_capacity()'s implementation into impl UpdateArgs block (similar to insert_key_value()) but it was actually impossible because the logic of this method needs reference to self object of the struct for which we need the trait to be implemented and then I just find as a solution to duplicate the code... Thank you for your advice! I moved the implementation into impl UpdateArgs for String and then just called the methods from there.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -219,49 +159,20 @@ impl Cmdline {
-        self.line.as_str()
-    }
+/// An abstraction for the boot arguments of a kernel command line.
+#[derive(Clone, Debug)]
```
We should follow the API guidelines and derive the other useful traits as well. Here are the Rust guidelines: https://rust-lang.github.io/api-guidelines/interoperability.html

andreitraistaru replied:
Adding implementation for Eq, PartialEq, Ord, PartialOrd, Hash, Default and Dsiplay traits as well for both BootArgs and InitArgs. Same traits where implemented for Cmdline, most of them using #derive, and a few of them by providing actual implementation.

andreeaflorescu commented on src/cmdline/mod.rs:
````diff
@@ -336,172 +250,641 @@ impl Cmdline {
+    /// let cl = Cmdline::new(100);
+    /// ```
+    /// [`Cmdline`]: struct.Cmdline.html
+    pub fn new(cmdline_capacity: usize) -> Result<Cmdline> {
````
With the current design of the API I don't think that a `cmdline_capacity` makes a lot of sense anymore. If we want to, we can give users the possibility to override the `MAX_CAPACITY` by having 2 contructors:
- one that takes a capacity as a parameter
- and one that doesn't take a capacity as a parameter, and just uses the defined DEFAULT instead.

andreitraistaru replied:
Actually, my concern here is that if we keep the capacity parameter of the Cmdline constructor and change the default MAX_CAPACITY value that we use when calling has_capacity() method of the UpdateArgs trait, we still don't handle this aspect very well: because we allow appending to the boot args string until reaching MAX_CAPACITY and also allow appending adding init args until reaches another MAX_CAPACITY in the init args string. When, eventually, call as_string() we throw an error because the joined cmdline exceeds MAX_CAPACITY; the best way to handle this case would be that in has_capacity() method to check somehow the length of both boot args string and init args string but implementing something like this means keeping a reference to init args in the BootArgs struct and vice versa which would introduce unjustified complexity from my point of view. Let's remove cmdline_capacity param from cmdline constructor and just keep a default MAX_CAPACITY for checks before inserts (preventing abuse) and before as_str as long as exceeding MAX_CAPACITY would definitely result into a erroneous cmdline; I chose 4096 as MAX_CAPACITY as long as this is the max size of a valid kernel cmdline according to the [spec](https://www.kernel.org/doc/html/v4.14/admin-guide/kernel-parameters.html)

andreitraistaru replied:
Ok, reading the comments below, I'll change the code and remove cmdline_capacity from constructor.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -336,172 +250,641 @@ impl Cmdline {
+
+        if self.init_args.0.is_empty() {
+            if self.boot_args.0.len() + 1 > self.cmdline_capacity {
+                // need space for the null terminator
```
In the documentation comment we're saying that we're not adding the nul terminator, which one is the correct comment?

andreitraistaru replied:
We don't add any '\0' char per se when returning the String from as_str() method. However, we need to consider the space required for the nul terminator within the cmdline_capacity so that's why I wrote the if condition adding + 1 (and added a comment to be more clear) - actually it's the same condition used as before, but written in another way.

andreitraistaru replied:
I'll come back to the old condition and remove the comment.

andreitraistaru replied:
But your questions triggered another question in my mind: who should be responsible for adding the '\0' char to the cmdline ? I'm asking because from what I understand, currently, nobody does this... I checked the linux-loader/loader crate and the function load_cmdline() does not do this; it just takes the cmdline.as_str() turns it into bytes and then writes these bytes into guest's memory; is my understanding right? shouldn't we add a '\0' at the end in as_str() method to make sure the cmdline is nul terminated?

andreeaflorescu replied:
We should be using CString to get a null terminated string.

andreeaflorescu reviewed (commented):
Part 1. Still need to look into try_from.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -336,172 +250,669 @@ impl Cmdline {
+    type Error = String;
+
+    fn try_from(
+        (cmdline_capacity, cmdline_raw): (usize, String),
```
If we want to have a capacity as well then I think implementing `TryFrom` is no longer a good idea because we have to use a tuple, so it's no longer clear how to initialize the `Cmdline`. In case we want to keep the capacity, we should just have a function on `Cmdline` so that we can have as named parameters a `value` and a `capacity`. I would recommend thought to drop `capacity` altogether and just check (as you're already doing) that the capacity of the generated string is less than the maximum allowed one by the linux kernel.

andreitraistaru replied:
Ok, implementing this way. I think that the cmdline_capacity param came from the fact that each arch has different cmdline_max sizes; we can allow the maximum of these values (4096)

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -219,49 +159,20 @@ impl Cmdline {
+pub struct BootArgs(String);
 
+impl BootArgs {
+    /// here we need to add any other functions that are specific to the boot args.
```
This comment should be removed.

andreitraistaru replied:
ups...; done

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -283,10 +194,11 @@ impl Cmdline {
+    ///     .add_virtio_mmio_device(1 << 12, GuestAddress(0x1000), 5, Some(42))
     ///     .unwrap();
-    /// let cl_cstring = CString::new(cl).unwrap();
+    /// let cl_cstring = CString::new(cl.as_str().unwrap()).unwrap();
```
I don't think this needs to be a CString anymore.

andreeaflorescu replied:
On a second thought, that where we get the null terminated string from. The reason why we're using CString here is that one, it was just very poorly documented.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -336,172 +250,669 @@ impl Cmdline {
+        if !self.0.is_empty() {
+            // this is NOT the first arg so we need to add an extra space between existing
+            // args and the new args to be added
+            auxiliary_space += 1;
```
Shouldn't it be enough to initialize this to 1 here instead of adding 1?

andreitraistaru replied:
simplifying even further and removing mutability by refactoring into: let auxiliary_space = if self.is_empty() { 0 } else { 1 };

andreeaflorescu reviewed (commented):
Part 2, still need to look into a few corner cases.

andreitraistaru commented on src/cmdline/mod.rs:
```diff
@@ -336,172 +250,669 @@ impl Cmdline {
+            } else {
+                Ok(self.boot_args.0.to_string())
+            }
+        } else if self.boot_args.0.len() + self.init_args.0.len() + 5 > self.cmdline_capacity {
```
Done.

andreeaflorescu commented on src/cmdline/mod.rs:
````diff
@@ -336,172 +269,736 @@ impl Cmdline {
+    /// assert!(cl.init_args_mut().insert_str("bar").is_ok());
+    /// assert_eq!(cl.as_str().unwrap(), "foo -- bar");
+    /// ```
+    pub fn as_str(&self) -> Result<String> {
````
To have a null terminated string here we can return a `CString`.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -345,42 +290,216 @@ impl Cmdline {
+    fn try_from(cmdline_raw: String) -> result::Result<Self, Self::Error> {
+        // Checking for ' -- ' sequence is required as long as a sequence like 'foo-- bar' should not be
+        // handled as a boot param 'foo' and an init arg 'bar'
+        let (mut boot_args, mut init_args) = match cmdline_raw.split_once(" -- ") {
```
Can we have a define for " -- " instead? For example, we can call it INIT_ARG_DELIMITER. Then, we should be able to use INIT_ARG_DELIMITER.len() instead of magic numbers in the code below.

andreitraistaru replied:
We already have such a define: INIT_ARGS_SEPARATOR; I don't see any magic number below apart some variations from the ' -- ' separator such as '-- ' or ' --' that are used to catch some corner cases (but their only usecase is in try_from() so I considered that is easier to see the actual value instead of another define which is used in only one place). Looks like this comment is older and I forgot to close it or something... ping me if I miss something

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -345,42 +290,216 @@ impl Cmdline {
+        boot_args = boot_args.trim();
+        init_args = init_args.trim();
+
+        let mut more_space = 1; // required for null terminator of the cmdline
```
It would make the code more readable if we instead define a `cmdline_size` instead of `more_space`. Then, the code will read more like a story:
- cmdline size initially is boot_args.len() + 1;
- if in the cmdline we also have `init_args`, then, the cmdline_size += init_args.len() + INIT_ARGS_DELIMITER.len().
- if the cmdline size is bigger than CMDLINE_MAXSIZE, then we return an error

andreitraistaru replied:
updating that

andreitraistaru commented on src/cmdline/mod.rs:
```diff
@@ -220,57 +168,51 @@ impl Cmdline {
+}
+
+impl UpdateArgs for String {
+    fn has_capacity(&self, more: usize) -> bool {
```
Yes, that was what I was concerned in one of the comments above. From my point of view the solution to keep track of the size of the init args string while adding a new boot arg (and vice versa) would introduce a great complexity in the code that is not justified (I suppose we need to add a reference to the init string inside the boot arg structure and vice versa to be able to have the check properly implemented at this level, but the design would be hard to maintain).

However, my solution was to keep has_capacity from `Args` in order to prevent abuse (not to allow adding more that MAX_CAPACITY bytes into any of the init args string or boot args string because that would definitely result in an error when calling as_cstring()) and the real check already exist in the as_cstring().

Another option would be to have a method under `Cmdline` that takes two params (one specifying the new arguments that needs to be appended and the second one to specify the type of param that we try to add). This method can check if there is still space (according to all corner cases) because `Cmdline` has access to both init and boot arg strings and after that we call the corresponding insert method to the `init args` or `boot args` structs. In other words, to implement a wrapper over insert* methods in `Cmdline`. Doing like this mens we no longer need methods `boot_args_mut()` and `init_args_mut()` from `Cmdline`.

Would you like to go with this last approach?

andreeaflorescu commented on src/cmdline/mod.rs:
````diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
+    /// assert!(cl.insert_string("foo", CmdlineArgType::BootArg).is_ok());
+    /// assert_eq!(cl.set_capacity(3), Err(Error::InvalidCapacity));
+    /// ```
+    pub fn set_capacity(&mut self, new_capacity: usize) -> Result<()> {
````
What use case does this function serve? We can already set the capacity in the constructor. When would we need to update that?

andreitraistaru replied:
removing this as discussed

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
     }
 
-    /// Validates and inserts a key-value pair into this command line.
+    /// Validates and inserts a key=value pair as a new cmdline argument.
```
Why change key=value? The `key-value` is written like that on purpose as it represents a concept. In the documentation comments we should use concepts, not implementation details as a general rule.

andreitraistaru replied:
keeping as it was initially

andreeaflorescu commented on src/cmdline/mod.rs:
````diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
     /// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foo=bar\0");
     /// ```
-    pub fn insert<T: AsRef<str>>(&mut self, key: T, val: T) -> Result<()> {
+    pub fn insert_key_value<T: AsRef<str>>(
````
Why not leave it as insert? The fact that it's key-value is clear from the parameters that it receives.

andreitraistaru replied:
going back to the old naming

andreeaflorescu commented on src/cmdline/mod.rs:
````diff
@@ -220,55 +402,41 @@ impl Cmdline {
     /// ```
-    pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
-        let s = slug.as_ref();
+    pub fn insert_string<T: AsRef<str>>(
````
Why change the name of this? It still receives something that can be a `str`, so the name was correct before.

andreitraistaru replied:
going back to the old naming

andreeaflorescu reviewed (commented):
Let's chat offline more about how we can get this PR to a mergable state.

lauralt commented on src/cmdline/mod.rs:
```diff
@@ -234,24 +218,95 @@ impl Cmdline {
+        let mut cmdline_size = self.get_null_terminated_representation_size();
+
+        // Checking if one extra space needs to be added before insertion of a new boot arg
+        cmdline_size += if !self.boot_args.is_empty() { 1 } else { 0 };
```
I see some unchecked operations in this PR, is it possible to have overflows? If not, would be nice to add comments about that.

andreitraistaru replied:
fixing that...

lauralt commented on src/cmdline/mod.rs:
```diff
@@ -234,24 +218,95 @@ impl Cmdline {
+        // Checking if one extra space needs to be added before insertion of a new boot arg
+        cmdline_size += if !self.boot_args.is_empty() { 1 } else { 0 };
+
+        // Adding extra size required for the insertion of `more` bytes as boot args
```
Is `more` a reference to some parameter or?

andreitraistaru replied:
it was just an outdated comment; thank you for notification!

andreitraistaru commented:
I added new tests for the try_from() functionality and fixed some issue discovered meanwhile. This is one of the most error-prone sections of the code, as long as there are many corner cases hard to detect

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -234,24 +218,117 @@ impl Cmdline {
+            };
+
+        // Adding extra size required for the insertion of the new boot args
+        cmdline_size = match cmdline_size.checked_add(s.len()) {
```
We should convert Options into Results for neat code:
```rust
cmdline_size = cmdline_size.checked_add(s.len()).ok_or(Error::TooLarge)?;
```

andreitraistaru replied:
done

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -234,24 +218,117 @@ impl Cmdline {
+
+        // Checking if one extra space needs to be added before insertion of a new boot arg
+        cmdline_size =
+            match cmdline_size.checked_add(if !self.boot_args.is_empty() { 1 } else { 0 }) {
```
We should have the if first becasue we shouldn't have to add 0 to size. We can also use `ok_or` to return an error in case of an overflow.

andreitraistaru replied:
done

aghecenco commented on src/cmdline/mod.rs:
````diff
@@ -234,24 +221,110 @@ impl Cmdline {
     /// ```
     pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
-        let s = slug.as_ref();
+        // Step 1: Check if the string provided is a valid boot arg string and remove any
````
These two `insert`s are really similar so I think you can comfortably extract the common bits into (yet another) helper - the flow is basically:
1. check that `slug` fits
2. put `slug` in either `boot_args` or `init_args`

So you could compress step 1. into a checker and just pass it how many extra chars it needs to accommodate, something like

```rust
fn check_size<T: AsRef<str>>(&mut self, slug: &T, num_extra_chars: usize) -> Result<()> {
        // Step 1: Check if the string provided is a valid boot arg string and remove any
        // leading or trailing whitespaces.
        let s = slug.as_ref().trim();
        valid_str(s)?;

        // Step 2: Check if cmdline capacity is not exceeded when inserting the arg
        // string provided.
        let cmdline_size = self.get_null_terminated_representation_size()
        .checked_add(num_extra_chars).ok_or(Error::TooLarge)?
        .checked_add(s.len()).ok_or(Error::TooLarge)?;

        if cmdline_size > self.capacity {
            Err(Error::TooLarge)
        } else {
            Ok(())
        }
    }
```

Then call it like this -

```rust
pub fn insert_str<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
        self.check_size(&slug, if self.boot_args.is_empty() { 0 } else { 1 })?;

        if !self.boot_args.is_empty() {
        // etc
    }

pub fn insert_init_args<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
        self.check_size(&slug, if self.init_args.is_empty() {
            INIT_ARGS_SEPARATOR.len()
        } else {
            1
        })?;

        if !self.init_args.is_empty() {
        // etc
}
````

aghecenco commented on src/cmdline/mod.rs:
```diff
@@ -343,17 +431,106 @@ impl Cmdline {
+
+        // Step 2: Check if capacity provided for the cmdline is not exceeded and create a new `Cmdline`
+        // if size check passes.
+        let mut cmdline_size = boot_args.len().checked_add(1).ok_or(Error::TooLarge)?;
+
+        if !init_args.is_empty() {
+            cmdline_size = cmdline_size
+                .checked_add(INIT_ARGS_SEPARATOR.len())
+                .ok_or(Error::TooLarge)?;
+
+            cmdline_size = cmdline_size
+                .checked_add(init_args.len())
+                .ok_or(Error::TooLarge)?;
```
nit: these `checked_add`s can be chained into something like

```rust
    let cmdline_size = boot_args
            .len()
            .checked_add(1)
            .and_then(|s| {
                if !init_args.is_empty() {
                    s.checked_add(INIT_ARGS_SEPARATOR.len() + init_args.len())
                } else {
                    Some(s)
                }
            })
            .ok_or(Error::TooLarge)?;
```

andreitraistaru replied:
As discussed, I'll let as it is, as long as we still need to have checked_add() for `INIT_ARGS_SEPARATOR.len() + init_args.len()`

aghecenco reviewed (commented):
I think the only thing missing at this point is the changelog update about the breaking changes, functionality-wise LGTM.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -223,7 +209,8 @@ impl Cmdline {
     }
 
-    /// Validates and inserts a string to the end of the current command line.
+    /// Validates and inserts a string representing boot args
```
I think this doc comment is a bit misleading. The purpose of the function is to "insert a string in the boot args". And then we can say: "Returns an error if the string is invalid". This was wrong before as well, but since we are updating it, we can also correct it. With the current wording it looks like the function does more than one thing which is not really the case.

andreeaflorescu commented on src/cmdline/mod.rs:
```diff
@@ -343,17 +431,106 @@ impl Cmdline {
-        cmdline.line.into_bytes()
+        match cmdline.as_cstring() {
+            Ok(cmdline_cstring) => cmdline_cstring.into_bytes_with_nul(),
+            Err(_) => Vec::from([0u8]),
```
This is a silent failure. We should change this to try_from instead and return the error from `into_bytes_with_nul` instead.

andreitraistaru replied:
solved

andreitraistaru commented on src/cmdline/mod.rs:
```diff
@@ -8,24 +8,21 @@
 pub enum Error {
     /// Operation would have resulted in a non-printable ASCII character.
     InvalidAscii,
-    /// Invalid device passed to the kernel command line builder.
```
done

andreeaflorescu commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+  than required for arguments (using `try_from`).
+- Added `insert_init_args` method allowing insertion of init arguments into `Cmdline`.
+
+## Changed
```
We also return a `Result` in `new` and check for invalid command line size.

andreeaflorescu replied:
And we also changed the implementation that converts a command line to a Vec<u8> and added error handling (i.e replaced From<Cmdline> with TryFrom<Cmdline>

andreitraistaru replied:
done

andreeaflorescu commented on CHANGELOG.md:
```diff
@@ -1,3 +1,18 @@
+- Added `InvalidCapacity` error type used when trying to create a `Cmdline`
+  either with zero capacity (using `new` or `try_from`) or with smaller capacity
+  than required for arguments (using `try_from`).
+- Added `insert_init_args` method allowing insertion of init arguments into `Cmdline`.
```
We should also say we added `try_from` that has support for both init args and command line args.

aghecenco reviewed (approved):
I am OK with the current changes, the changelog update is maybe a bit too verbose but that can be adjusted when we release.
